diff --git a/dlls/ntdll/unix/server.c b/dlls/ntdll/unix/server.c
index b18ae5b9726..3daddf924fd 100644
--- a/dlls/ntdll/unix/server.c
+++ b/dlls/ntdll/unix/server.c
@@ -1402,10 +1402,17 @@ static int setup_config_dir(void)
 
     if (!mkdir( "dosdevices", 0777 ))
     {
+#ifdef __ANDROID__
+        mkdir( "drive_d", 0777 );
+        symlink( "../drive_c", "dosdevices/c:" );
+        symlink( "/storage/emulated/0/", "dosdevices/d:" );
+        symlink( "/data/data/app.gamenative/files/imagefs/", "dosdevices/z:" );
+#else
         mkdir( "drive_c", 0777 );
         set_case_insensitive( "drive_c" );
         symlink( "../drive_c", "dosdevices/c:" );
         symlink( "/", "dosdevices/z:" );
+#endif
     }
     else if (errno != EEXIST) fatal_perror( "cannot create %s/dosdevices", config_dir );
 
@@ -1771,6 +1778,11 @@ size_t server_init_process(void)
                 inproc_device_fd = FSYNC_USED_BY_SERVER;
                 fsync_init( pid );
             }
+            else if (reply->inproc_device == ESYNC_USED_BY_SERVER)
+            {
+                int shm_fd = wine_server_receive_fd( &handle );
+                esync_init( shm_fd );
+            }
             else if (reply->inproc_device)
             {
                 inproc_device_fd = wine_server_receive_fd( &handle );
