diff --git a/dlls/wow64/virtual.c b/dlls/wow64/virtual.c
--- a/dlls/wow64/virtual.c
+++ b/dlls/wow64/virtual.c
@@ -694,11 +694,18 @@
     }
 
     case MemoryWineUnixWow64Funcs:
+    case MemoryWineLoadUnixLibByNameWow64:
         return STATUS_INVALID_INFO_CLASS;
 
     case MemoryWineUnixFuncs:
         status = NtQueryVirtualMemory( handle, addr, MemoryWineUnixWow64Funcs, ptr, len, &res_len );
         break;
+    case MemoryWineLoadUnixLibByName:
+        status = NtQueryVirtualMemory( handle, addr, MemoryWineLoadUnixLibByNameWow64, ptr, len, &res_len );
+        break;
+    case MemoryWineUnloadUnixLib:
+        status = NtQueryVirtualMemory( handle, addr, class, ptr, len, &res_len );
+        break;
 
     default:
         FIXME( "unsupported class %u\n", class );
