diff --git a/dlls/ntdll/unix/server.c b/dlls/ntdll/unix/server.c
index 1629ee63e6..ef77dfc6f2 100644
--- a/dlls/ntdll/unix/server.c
+++ b/dlls/ntdll/unix/server.c
@@ -76,6 +76,8 @@
 #include "winnt.h"
 #include "winioctl.h"
 #include "wine/server.h"
+#include "fsync.h"
+#include "esync.h"
 #include "wine/debug.h"
 #include "unix_private.h"
 #include "ddk/wdm.h"
@@ -1694,7 +1696,17 @@ size_t server_init_process(void)
             info_size         = reply->info_size;
             server_start_time = reply->server_start;
             supported_machines_count = wine_server_reply_size( reply ) / sizeof(*supported_machines);
-            if (reply->inproc_device)
+            if (reply->inproc_device == FSYNC_USED_BY_SERVER)
+            {
+                inproc_device_fd = FSYNC_USED_BY_SERVER;
+                fsync_init( pid );
+            }
+            else if (reply->inproc_device == ESYNC_USED_BY_SERVER)
+            {
+                inproc_device_fd = ESYNC_USED_BY_SERVER;
+                esync_init( wine_server_receive_fd( &handle ) );
+            }
+            else if (reply->inproc_device)
             {
                 inproc_device_fd = wine_server_receive_fd( &handle );
                 assert( handle == reply->inproc_device );
