diff --git a/server/inproc_sync.c b/server/inproc_sync.c
index 522d9a4e32..f5d27e9382 100644
--- a/server/inproc_sync.c
+++ b/server/inproc_sync.c
@@ -33,6 +33,12 @@
 #include "thread.h"
 #include "user.h"
 
+#include "esync.h"
+#include "fsync.h"
+#ifdef HAVE_SYS_EVENTFD_H
+#include <sys/eventfd.h>
+#endif
+
 #ifdef HAVE_LINUX_NTSYNC_H
 # include <linux/ntsync.h>
 #endif
@@ -47,6 +53,8 @@
 int get_inproc_device_fd(void)
 {
     static int fd = -2;
+    if (do_fsync()) return FSYNC_USED_BY_SERVER;
+    if (do_esync()) return ESYNC_USED_BY_SERVER;
     if (fd == -2) fd = open( "/dev/ntsync", O_CLOEXEC | O_RDONLY );
     return fd;
 }
@@ -56,6 +64,7 @@ struct inproc_sync
     struct object          obj;  /* object header */
     enum inproc_sync_type  type;
     int                    fd;
+    unsigned int           esync_shm_idx;  /* ESYNC SHM INDEX */
     struct list            entry;
 };
 
@@ -86,8 +95,23 @@ struct inproc_sync *create_inproc_internal_sync( int manual, int signaled )
     struct inproc_sync *event;
 
     if (!(event = alloc_object( &inproc_sync_ops ))) return NULL;
-    event->type = INPROC_SYNC_INTERNAL;
-    event->fd   = ioctl( get_inproc_device_fd(), NTSYNC_IOC_CREATE_EVENT, &args );
+    if (do_fsync())
+    {
+        event->type = manual ? FSYNC_MANUAL_SERVER : FSYNC_AUTO_SERVER;
+        event->fd = fsync_alloc_shm( signaled, 0xdeadbeef );
+    }
+    else if (do_esync())
+    {
+        event->type = manual ? ESYNC_MANUAL_SERVER : ESYNC_AUTO_SERVER;
+        event->fd = eventfd( signaled ? 1 : 0, EFD_CLOEXEC | EFD_NONBLOCK );
+        if (event->fd != -1)
+            event->esync_shm_idx = esync_alloc_shm( event->fd, event->type, signaled, 0 );
+    }
+    else
+    {
+        event->type = INPROC_SYNC_INTERNAL;
+        event->fd   = ioctl( get_inproc_device_fd(), NTSYNC_IOC_CREATE_EVENT, &args );
+    }
     list_init( &event->entry );
 
     if (event->fd == -1)
@@ -105,8 +129,23 @@ struct inproc_sync *create_inproc_event_sync( int manual, int signaled )
     struct inproc_sync *event;
 
     if (!(event = alloc_object( &inproc_sync_ops ))) return NULL;
-    event->type = INPROC_SYNC_EVENT;
-    event->fd   = ioctl( get_inproc_device_fd(), NTSYNC_IOC_CREATE_EVENT, &args );
+    if (do_fsync())
+    {
+        event->type = manual ? FSYNC_MANUAL_EVENT : FSYNC_AUTO_EVENT;
+        event->fd = fsync_alloc_shm( signaled, 0xdeadbeef );
+    }
+    else if (do_esync())
+    {
+        event->type = manual ? ESYNC_MANUAL_EVENT : ESYNC_AUTO_EVENT;
+        event->fd = eventfd( signaled ? 1 : 0, EFD_CLOEXEC | EFD_NONBLOCK );
+        if (event->fd != -1)
+            event->esync_shm_idx = esync_alloc_shm( event->fd, event->type, signaled, 0 );
+    }
+    else
+    {
+        event->type = INPROC_SYNC_EVENT;
+        event->fd   = ioctl( get_inproc_device_fd(), NTSYNC_IOC_CREATE_EVENT, &args );
+    }
     list_init( &event->entry );
 
     if (event->fd == -1)
@@ -124,8 +163,23 @@ struct inproc_sync *create_inproc_mutex_sync( thread_id_t owner, unsigned int co
     struct inproc_sync *mutex;
 
     if (!(mutex = alloc_object( &inproc_sync_ops ))) return NULL;
-    mutex->type = INPROC_SYNC_MUTEX;
-    mutex->fd   = ioctl( get_inproc_device_fd(), NTSYNC_IOC_CREATE_MUTEX, &args );
+    if (do_fsync())
+    {
+        mutex->type = FSYNC_MUTEX;
+        mutex->fd = fsync_alloc_shm( owner, count );
+    }
+    else if (do_esync())
+    {
+        mutex->type = ESYNC_MUTEX;
+        mutex->fd = eventfd( owner == 0 ? 1 : 0, EFD_CLOEXEC | EFD_NONBLOCK );
+        if (mutex->fd != -1)
+            mutex->esync_shm_idx = esync_alloc_shm( mutex->fd, mutex->type, owner == 0 ? 1 : 0, 0 );
+    }
+    else
+    {
+        mutex->type = INPROC_SYNC_MUTEX;
+        mutex->fd   = ioctl( get_inproc_device_fd(), NTSYNC_IOC_CREATE_MUTEX, &args );
+    }
     list_add_tail( &inproc_mutexes, &mutex->entry );
 
     if (mutex->fd == -1)
@@ -143,8 +197,23 @@ struct inproc_sync *create_inproc_semaphore_sync( unsigned int initial, unsigned
     struct inproc_sync *sem;
 
     if (!(sem = alloc_object( &inproc_sync_ops ))) return NULL;
-    sem->type = INPROC_SYNC_SEMAPHORE;
-    sem->fd   = ioctl( get_inproc_device_fd(), NTSYNC_IOC_CREATE_SEM, &args );
+    if (do_fsync())
+    {
+        sem->type = FSYNC_SEMAPHORE;
+        sem->fd = fsync_alloc_shm( initial, max );
+    }
+    else if (do_esync())
+    {
+        sem->type = ESYNC_SEMAPHORE;
+        sem->fd = eventfd( initial, EFD_CLOEXEC | EFD_NONBLOCK | EFD_SEMAPHORE );
+        if (sem->fd != -1)
+            sem->esync_shm_idx = esync_alloc_shm( sem->fd, sem->type, initial, max );
+    }
+    else
+    {
+        sem->type = INPROC_SYNC_SEMAPHORE;
+        sem->fd   = ioctl( get_inproc_device_fd(), NTSYNC_IOC_CREATE_SEM, &args );
+    }
     list_init( &sem->entry );
 
     if (sem->fd == -1)
@@ -167,14 +236,18 @@ void signal_inproc_sync( struct inproc_sync *sync )
 {
     __u32 count;
     if (debug_level) fprintf( stderr, "set_inproc_event %d\n", sync->fd );
-    ioctl( sync->fd, NTSYNC_IOC_EVENT_SET, &count );
+    if (do_fsync()) fsync_set_event( sync->fd );
+    else if (do_esync()) esync_set_event( sync->fd, sync->esync_shm_idx, sync->type );
+    else            ioctl( sync->fd, NTSYNC_IOC_EVENT_SET, &count );
 }
 
 void reset_inproc_sync( struct inproc_sync *sync )
 {
     __u32 count;
     if (debug_level) fprintf( stderr, "reset_inproc_event %d\n", sync->fd );
-    ioctl( sync->fd, NTSYNC_IOC_EVENT_RESET, &count );
+    if (do_fsync()) fsync_reset_event( sync->fd );
+    else if (do_esync()) esync_reset_event( sync->fd, sync->esync_shm_idx, sync->type );
+    else            ioctl( sync->fd, NTSYNC_IOC_EVENT_RESET, &count );
 }
 
 static int inproc_sync_signal( struct object *obj, unsigned int access, int signal )
@@ -182,7 +255,8 @@ static int inproc_sync_signal( struct object *obj, unsigned int access, int sign
     struct inproc_sync *sync = (struct inproc_sync *)obj;
     assert( obj->ops == &inproc_sync_ops );
 
-    assert( sync->type == INPROC_SYNC_INTERNAL || sync->type == INPROC_SYNC_EVENT ); /* never called for mutex / semaphore */
+    if (!do_fsync() && !do_esync())
+        assert( sync->type == INPROC_SYNC_INTERNAL || sync->type == INPROC_SYNC_EVENT ); /* never called for mutex / semaphore */
     assert( signal == 0 || signal == 1 ); /* never called from signal_object */
 
     if (signal) signal_inproc_sync( sync );
@@ -202,11 +276,24 @@ void abandon_inproc_mutexes( thread_id_t tid )
 {
     struct inproc_sync *mutex;
 
+    if (do_fsync())
+    {
+        LIST_FOR_EACH_ENTRY( mutex, &inproc_mutexes, struct inproc_sync, entry )
+            fsync_abandon_mutex( mutex->fd, tid );
+        return;
+    }
+    else if (do_esync())
+    {
+        LIST_FOR_EACH_ENTRY( mutex, &inproc_mutexes, struct inproc_sync, entry )
+            esync_abandon_mutex( mutex->fd, mutex->esync_shm_idx, tid );
+        return;
+    }
+
     LIST_FOR_EACH_ENTRY( mutex, &inproc_mutexes, struct inproc_sync, entry )
         ioctl( mutex->fd, NTSYNC_IOC_MUTEX_KILL, &tid );
 }
 
-static int get_obj_inproc_sync( struct object *obj, int *type )
+static int get_obj_inproc_sync( struct object *obj, int *type, unsigned int *esync_shm_idx )
 {
     struct object *sync;
     int fd = -1;
@@ -217,6 +304,7 @@ static int get_obj_inproc_sync( struct object *obj, int *type )
         struct inproc_sync *inproc = (struct inproc_sync *)sync;
         *type = inproc->type;
         fd = inproc->fd;
+        if (esync_shm_idx) *esync_shm_idx = inproc->esync_shm_idx;
     }
 
     release_object( sync );
@@ -267,7 +355,7 @@ void abandon_inproc_mutexes( thread_id_t tid )
 {
 }
 
-static int get_obj_inproc_sync( struct object *obj, int *type )
+static int get_obj_inproc_sync( struct object *obj, int *type, unsigned int *esync_shm_idx )
 {
     return -1;
 }
@@ -278,12 +366,19 @@ DECL_HANDLER(get_inproc_sync_fd)
 {
     struct object *obj;
     int fd;
+    unsigned int esync_shm_idx = 0;
 
     if (!(obj = get_handle_obj( current->process, req->handle, 0, NULL ))) return;
 
     reply->access = get_handle_access( current->process, req->handle );
 
-    if ((fd = get_obj_inproc_sync( obj, &reply->type )) < 0) set_error( STATUS_NOT_IMPLEMENTED );
+    if ((fd = get_obj_inproc_sync( obj, &reply->type, &esync_shm_idx )) < 0) set_error( STATUS_NOT_IMPLEMENTED );
+    else if (do_fsync()) reply->sync_shm_idx = fsync_grab_shm_idx( fd );
+    else if (do_esync())
+    {
+        reply->sync_shm_idx = esync_shm_idx;
+        send_client_fd( current->process, fd, req->handle );
+    }
     else send_client_fd( current->process, fd, req->handle );
 
     release_object( obj );
