diff --git a/server/thread.c b/server/thread.c
index bdf48c4bf8..446cc79288 100644
--- a/server/thread.c
+++ b/server/thread.c
@@ -60,6 +60,8 @@
 #include "user.h"
 #include "security.h"
 
+#include "fsync.h"
+
 
 /* thread queues */
 
@@ -1743,7 +1745,11 @@ DECL_HANDLER(init_first_thread)
     set_reply_data( supported_machines,
                     min( supported_machines_count * sizeof(unsigned short), get_reply_max_size() ));
 
-    if ((fd = get_inproc_device_fd()) >= 0)
+    if (do_fsync())
+    {
+        reply->inproc_device = FSYNC_USED_BY_SERVER;
+    }
+    else if ((fd = get_inproc_device_fd()) >= 0)
     {
         reply->inproc_device = get_process_id( process ) | 1;
         send_client_fd( process, fd, reply->inproc_device );
@@ -2380,6 +2386,7 @@ DECL_HANDLER(get_inproc_alert_fd)
     int fd;
 
     if ((fd = get_inproc_sync_fd( current->alert_sync )) < 0) set_error( STATUS_INVALID_PARAMETER );
+    else if (do_fsync()) reply->sync_shm_idx = fd;
     else
     {
         reply->handle = get_thread_id( current ) | 1; /* arbitrary token */
