diff --git a/dlls/winepulse.drv/pulse.c b/dlls/winepulse.drv/pulse.c
index ef13a1ed0bb..1ad29870a25 100644
--- a/dlls/winepulse.drv/pulse.c
+++ b/dlls/winepulse.drv/pulse.c
@@ -243,8 +243,10 @@ static NTSTATUS pulse_process_attach(void *args)
     pthread_mutexattr_t attr;
 
     pthread_mutexattr_init(&attr);
+#ifndef __ANDROID__
     pthread_mutexattr_setprotocol(&attr, PTHREAD_PRIO_INHERIT);
     pthread_mutexattr_setrobust(&attr, PTHREAD_MUTEX_ROBUST);
+#endif
 
     if (pthread_mutex_init(&pulse_mutex, &attr) != 0)
         pthread_mutex_init(&pulse_mutex, NULL);
@@ -282,9 +284,13 @@ static void pulse_main_loop(void *args)
     pulse_ml = pa_mainloop_new();
     pa_mainloop_set_poll_func(pulse_ml, pulse_poll_func, NULL);
     NtSetEvent(event, NULL);
+#ifdef __ANDROID__
+    pa_mainloop_run(pulse_ml, &ret);
+#else
     pthread_cleanup_push(pulse_main_loop_thread_cleanup, NULL);
     pa_mainloop_run(pulse_ml, &ret);
     pthread_cleanup_pop(0);
+#endif
     pa_mainloop_free(pulse_ml);
     pulse_unlock();
     PsTerminateSystemThread( 0 );
